# Задание практической работы

## 1. Создание и изменение массивов

Создайте массив NumPy размером 3×3, содержащий числа от 1 до 9. Умножьте все элементы массива на 2, а затем замените все элементы больше 10 на 0. Выведите итоговый массив.

In [1]:
import numpy as np
arr = np.arange(1, 10).reshape(3, 3)
print("Исходный массив:\n", arr)
arr *= 2
arr[arr > 10] = 0
print("\nИтоговый массив:\n", arr)

Исходный массив:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]

Итоговый массив:
 [[ 2  4  6]
 [ 8 10  0]
 [ 0  0  0]]


## 2. Работа с булевыми масками

Создайте массив NumPy из 20 случайных целых чисел от 1 до 100. Найдите и выведите все элементы, которые делятся на 5 без остатка. Затем замените их на -1 и выведите обновленный массив.

In [2]:
import numpy as np
arr = np.random.randint(1, 101, 20)
print("Исходный массив:", arr)
div_by_5 = arr[arr % 5 == 0]
print("Элементы, делящиеся на 5:", div_by_5)
arr[arr % 5 == 0] = -1
print("Обновленный массив:", arr)

Исходный массив: [51 37 11 57 66 78 19 39 99 14  7 85  3 40 43 30 73 79 53 26]
Элементы, делящиеся на 5: [85 40 30]
Обновленный массив: [51 37 11 57 66 78 19 39 99 14  7 -1  3 -1 43 -1 73 79 53 26]


## 3. Объединение и разбиение массивов

Создайте два массива NumPy размером 1×5, заполненные случайными числами от 0 до 50. 
1. Объедините эти массивы в один двумерный массив (по строкам). 
2. Разделите полученный массив на два массива, каждый из которых содержит 5 элементов. 
Выведите все промежуточные и итоговые результаты.

In [3]:
import numpy as np
arr1 = np.random.randint(0, 51, 5)
arr2 = np.random.randint(0, 51, 5)
print("Первый массив:", arr1)
print("Второй массив:", arr2)
comb = np.vstack((arr1, arr2))
print("\nОбъединенный массив:\n", comb)
sp_arr1, sp_arr2 = np.vsplit(comb, 2)
print("\nПервый разделенный массив:", sp_arr1)
print("Второй разделенный массив:", sp_arr2)

Первый массив: [24 24 10 43 19]
Второй массив: [44 25 43 46 30]

Объединенный массив:
 [[24 24 10 43 19]
 [44 25 43 46 30]]

Первый разделенный массив: [[24 24 10 43 19]]
Второй разделенный массив: [[44 25 43 46 30]]


## 4. Генерация и работа с линеными последовательностями

Создайте массив из 50 чисел, равномерно распределенных от -10 до 10. Вычислите сумму всех элементов, сумму положительных элементов и сумму отрицательных элементов. Выведите результаты.

In [6]:
import numpy as np
arr = np.linspace(-10, 10, 50)
print("Массив:", arr)
total_sum = np.sum(arr)
pos_sum = np.sum(arr[arr > 0])
neg_sum = np.sum(arr[arr < 0])
print("\nОбщая сумма всех элементов:", total_sum)
print("Сумма положительных элементов:", pos_sum)
print("Сумма отрицательных элементов:", neg_sum)

Массив: [-10.          -9.59183673  -9.18367347  -8.7755102   -8.36734694
  -7.95918367  -7.55102041  -7.14285714  -6.73469388  -6.32653061
  -5.91836735  -5.51020408  -5.10204082  -4.69387755  -4.28571429
  -3.87755102  -3.46938776  -3.06122449  -2.65306122  -2.24489796
  -1.83673469  -1.42857143  -1.02040816  -0.6122449   -0.20408163
   0.20408163   0.6122449    1.02040816   1.42857143   1.83673469
   2.24489796   2.65306122   3.06122449   3.46938776   3.87755102
   4.28571429   4.69387755   5.10204082   5.51020408   5.91836735
   6.32653061   6.73469388   7.14285714   7.55102041   7.95918367
   8.36734694   8.7755102    9.18367347   9.59183673  10.        ]

Общая сумма всех элементов: 7.105427357601002e-15
Сумма положительных элементов: 127.55102040816328
Сумма отрицательных элементов: -127.55102040816327


## 5. Работа с диагональными и единичными матрицами

Создайте: 
1. Единичную матрицу размером 4×4. 
2. Диагональную матрицу размером 4×4 с диагональными элементами [5, 10, 15, 20] (не использовать циклы).

Найдите сумму всех элементов каждой из этих матриц и сравните результаты.


In [7]:
import numpy as np
arr1 = np.eye(4)
print("Единичная матрица:\n", arr1)
arr2 = np.diag([5, 10, 15, 20])
print("\nДиагональная матрица:\n", arr2)
sum1 = np.sum(arr1)
sum2 = np.sum(arr2)
print("\nСумма элементов единичной матрицы:", sum1)
print("Сумма элементов диагональной матрицы:", sum2)

Единичная матрица:
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]

Диагональная матрица:
 [[ 5  0  0  0]
 [ 0 10  0  0]
 [ 0  0 15  0]
 [ 0  0  0 20]]

Сумма элементов единичной матрицы: 4.0
Сумма элементов диагональной матрицы: 50


## 6. Создание и базовые операции с матрицами

Создайте две квадратные матрицы NumPy размером 3×3, заполненные случайными целыми числами от 1 до 20. Вычислите и выведите: 
1. Их сумму
2. Их разность 
3. Их поэлементное произведение


In [9]:
import numpy as np
matrix1 = np.random.randint(1, 21, (3, 3))
matrix2 = np.random.randint(1, 21, (3, 3))
print("Матрица 1:\n", matrix1)
print("\nМатрица 2:\n", matrix2)
matrix_sum = matrix1 + matrix2
print("\nСумма матриц:\n", matrix_sum)
matrix_diff = matrix1 - matrix2
print("\nРазность матриц:\n", matrix_diff)
matrix_prod = matrix1 * matrix2
print("\nПоэлементное произведение:\n", matrix_prod)

Матрица 1:
 [[ 3  3 20]
 [14 12 10]
 [15 18 15]]

Матрица 2:
 [[14 10 15]
 [ 3 16 18]
 [20  9  9]]

Сумма матриц:
 [[17 13 35]
 [17 28 28]
 [35 27 24]]

Разность матриц:
 [[-11  -7   5]
 [ 11  -4  -8]
 [ -5   9   6]]

Поэлементное произведение:
 [[ 42  30 300]
 [ 42 192 180]
 [300 162 135]]


## 7. Умножение матриц

Создайте две матрицы NumPy: 
1. Первую размером 2×3, заполненную случайными числами от 1 до 10. 
2. Вторую размером 3×2, заполненную случайными числами от 1 до 10.
   
Выполните матричное умножение (@ или np.dot) и выведите результат


In [11]:
import numpy as np
matrix1 = np.random.randint(1, 11, (2, 3))
print("Первая матрица:\n", matrix1)
matrix2 = np.random.randint(1, 11, (3, 2))
print("\nВторая матрица:\n", matrix2)
matrix_prod = matrix1 @ matrix2
print("\nРезультат матричного умножения:\n", matrix_prod)

Первая матрица:
 [[1 5 7]
 [4 9 2]]

Вторая матрица:
 [[ 6  1]
 [ 7  3]
 [10  1]]

Результат матричного умножения:
 [[111  23]
 [107  33]]


## 8. Определитель и обратная матрица

Создайте случайную квадратную матрицу 3×3. Найдите и выведите: 
1. Определитель этой матрицы 
2. Обратную матрицу (если существует, иначе выведите сообщение, что матрица вырождена)
   
Используйте функции np.linalg.det и np.linalg.inv.


In [13]:
import numpy as np
matrix = np.random.randint(1, 10, (3, 3))
determ = np.linalg.det(matrix)
if np.isclose(determ, 0):
    print("Матрица вырожденна")
else:
    print("Обратная матрица:\n", np.linalg.inv(matrix))

Обратная матрица:
 [[ 0.26666667  0.13333333 -0.63333333]
 [-0.03333333  0.23333333  0.01666667]
 [-0.1        -0.3         0.55      ]]


## 9. Транспонирование и след матрицы

Создайте матрицу NumPy размером 4×4, содержащую случайные целые числа от 1 до 50. Выведите: 
1. Исходную матрицу 
2. Транспонированную матрицу 
3. След матрицы (сумму элементов на главной диагонали)
   
Используйте np.trace для нахождения следа.


In [14]:
import numpy as np
matrix = np.random.randint(1, 51, (4, 4))
print("Исходная матрица:\n", matrix)
transposed_matrix = matrix.T
print("\nТранспонированная матрица:\n", transposed_matrix)
trace = np.trace(matrix)
print("\nСлед матрицы:", trace)

Исходная матрица:
 [[ 3 32 15  7]
 [23 21 18  6]
 [12  6 39 18]
 [22 32 16 14]]

Транспонированная матрица:
 [[ 3 23 12 22]
 [32 21  6 32]
 [15 18 39 16]
 [ 7  6 18 14]]

След матрицы: 77


## 10. Системы линейных уравнений

Решить систему линейных уравнений.

\begin{cases}
2x + 3y - z = 5 \\
4x - y + 2z = 6 \\
-3x + 5y + 4z = -2
\end{cases}

Используйте матричное представление Ax = B , где A – матрица коэффициентов, x – вектор неизвестных, B – вектор правой части. Решите систему с помощью np.linalg.solve и выведите результат.

In [15]:
import numpy as np
A = np.array([
    [2, 3, -1],
    [4, -1, 2],
    [-3, 5, 4]
])
B = np.array([5, 6, -2])
x = np.linalg.solve(A, B)
print("Решение системы уравнений:\n", x)

Решение системы уравнений:
 [1.63963964 0.57657658 0.00900901]


## 11. Индивидуальное задание

Каждое задание предусматривает построение системы линейных уравнений. Решите полученную систему уравнений с использованием библиотеки NumPy. Для решения системы используйте метод Крамера и матричный метод. Сравните полученные результаты, с результатами, полученными с помощью np.linalg.solve.

**№9. Отпимальный выбор топлива.** Автомобиль использует три вида топлива в разных пропорциях. Бензин даёт 30 МДж энергии на литр, дизель — 35 МДж, газ — 25 МДж. Для поездки требуется 100 МДж энергии. Водитель хочет использовать бензина в два раза больше, чем газа, а количество дизеля равно количеству газа. Сколько литров каждого топлива нужно залить?

Решение:
Пусть:

1. x - литры бензина
2. y - литры дизеля
3. z - литры газа

Система уравнений:

1. Энергия: 30x+35y+25x=100
2. Бензин/газ: x=2z или x-2z=0
3. Дизель/газ: y=z или y-z=0

In [16]:
import numpy as np
A = np.array([
    [30, 35, 25],
    [1, 0, -2],
    [0, 1, -1]
])
B = np.array([100, 0, 0])
print("Матрица коэффициентов A:")
print(A)
print("\nВектор правой части B:")
print(B)

Матрица коэффициентов A:
[[30 35 25]
 [ 1  0 -2]
 [ 0  1 -1]]

Вектор правой части B:
[100   0   0]


1. Метод Крамера

In [21]:
det_A = np.linalg.det(A)
A_x = A.copy()
A_x[:, 0] = B

A_y = A.copy()
A_y[:, 1] = B

A_z = A.copy()
A_z[:, 2] = B

det_Ax = np.linalg.det(A_x)
det_Ay = np.linalg.det(A_y) 
det_Az = np.linalg.det(A_z)

x_kramer = det_Ax / det_A
y_kramer = det_Ay / det_A
z_kramer = det_Az / det_A

print(f"Бензин: {x_kramer:.2f} литров")
print(f"Дизель: {y_kramer:.2f} литров")
print(f"Газ: {z_kramer:.2f} литров")

Бензин: 1.67 литров
Дизель: 0.83 литров
Газ: 0.83 литров


2. Матричный метод (A⁻¹B)

In [19]:
A_inv = np.linalg.inv(A)
sol_matrix = A_inv @ B
    
print(f"Бензин: {sol_matrix[0]:.2f} литров")
print(f"Дизель: {sol_matrix[1]:.2f} литров")
print(f"Газ: {sol_matrix[2]:.2f} литров")

Бензин: 1.67 литров
Дизель: 0.83 литров
Газ: 0.83 литров


3. Метод np.linalg.solve

In [18]:
sol_solve = np.linalg.solve(A, B)

print(f"Бензин: {sol_solve[0]:.2f} литров")
print(f"Дизель: {sol_solve[1]:.2f} литров") 
print(f"Газ: {sol_solve[2]:.2f} литров")

Бензин: 1.67 литров
Дизель: 0.83 литров
Газ: 0.83 литров


## Все методы дали одинаковый результат:
Бензин: 1.67 литров

Дизель: 0.83 литров

Газ: 0.83 литров